File: Road traffic data/C/stack.c
//NOM: CHEYCHOU MOUAFO JUNIOR   mat: 21T2374
#include <stdio.h>
#include <stdlib.h>
#define MAX 300 // la taille maximum du tableau qui vas contenir les données sur le trajet
// La structure de donnée qui vas contenir horodateur
typedef struct {
    int jour;
    int mois;
    int ans;
    int heure;
    int min;
    int sec;
}horodate;
// La structure de donnée pour la date
typedef struct {
    int jour;
    int mois;
    int ans;
}date;

typedef struct{
    horodate horodateur;
    char sexe;
    int age;
    date Date;
    char period[10];
    char repas[MAX];
    char unite[30];
    char lieu[25];
    int nombre;
}data;

typedef struct list{
    data val;
    struct list* next;
}list;
/* LA FONCTION CREER QUI SERT A CREER LE PREMIER ELEMENT DE LA LISTE, ET PREND A PARAMETRE LA DONNEE DE PREMIER ELEMENT*/
list *creer(data x){
    list *li;
    li = (list*)malloc(sizeof(list));
    li->val = x;
    li->next = NULL;
    return li;
}
/*LA FONCTION AJOUTER SERT A AJOUTER DES ELEMENTS A UNE LISTE DEJA EXISTANT*/
list *push(list *li, data val){
    list *ptr = li, *element;

    element = (list*)malloc(sizeof(list));
    element->val = val;
    element->next = li;

    return element;
}
/*LA FONCTION SAVE LIS LES DONNEES A STRUCTURER DU FICHIER "urban_data.txt" ET LES MET DANS UNE LISTE*/
 list *save_stack(){
     FILE *in = fopen("foodData.txt","r");
     int i,count = 1;
     data x;
     list *tete;
     char ch;

     if(in == NULL){
        printf("Erreur lors de l'ouverture du fichier !!! :(\n");
        return NULL;
     }
     fscanf(in,"%d",&x.horodateur.jour);
     if(feof(in))
        printf("Desolé, le fichier est vide :(\n");
     while(x.horodateur.jour != EOF){
        i = 0;
        if(feof(in))
            break;
        fscanf(in,"%d%d%d%d%d",&x.horodateur.mois,&x.horodateur.ans,&x.horodateur.heure,&x.horodateur.min,&x.horodateur.sec);
        x.sexe = getc(in);//recupere l'espace qu'il ya entre le champs seconde et sexe
        x.sexe = getc(in);//ecrase la valeur precedente(l'espace) avec celui du sexe
        fscanf(in,"%d%d%d%d%s",&x.age,&x.Date.jour,&x.Date.mois,&x.Date.ans,x.period);
        ch = getc(in);
        ch = getc(in);
        while((ch = getc(in))!='"'){
            x.repas[i++]=ch;
        }
        x.repas[i]='\0';//indique la fin de la chaine de caractère
        ch = getc(in);
        ch = getc(in);
        i = 0;
        while((ch = getc(in))!='"'){
            x.unite[i++]=ch;
        }
        x.unite[i]='\0';//indique la fin de la chaine de caractère
        fscanf(in,"%s%d",x.lieu,&x.nombre);
        if(count == 1){
            tete = creer(x);
        }else{
            tete = push(tete,x);
        }
        count++;
        fscanf(in,"%d",&x.horodateur.jour);

     }

     fclose(in);
    return tete;
 }
 //LA FONCION afficher SERT A AFFICHER LES ELEMENTS DE LA LISTE PASSER EN PARAMETRE
 void print_stack(list* li){// Elle affiche en FIFO (First In First Out)
    list* ptr = li;
    while(ptr!=NULL){
        printf("%d/%d/%-5d%d:%d:%-5d",ptr->val.horodateur.jour,ptr->val.horodateur.mois,ptr->val.horodateur.ans,ptr->val.horodateur.heure,ptr->val.horodateur.min,ptr->val.horodateur.sec);
        printf("%-3c%-3d%d/%d/%-7d",ptr->val.sexe,ptr->val.age,ptr->val.Date.jour,ptr->val.Date.mois,ptr->val.Date.ans);
        printf("%-10s%-65s%-20s%-15s%-5d\n",ptr->val.period,ptr->val.repas,ptr->val.unite,ptr->val.lieu,ptr->val.nombre);
        ptr = ptr->next;
     }
 }
 //LA FONCTION search PERMET DE RECHERCHER ET A AFFICHER LES ELEMENTS D'UNE LISTE QUI ONT UN AGE SPECIFIQUE
 void search(list *li,int age){
    list *ptr = li;
    int i=0;
    while(ptr != NULL){
        if(ptr->val.age == age){
            printf("%d/%d/%-5d%d:%d:%-5d",ptr->val.horodateur.jour,ptr->val.horodateur.mois,ptr->val.horodateur.ans,ptr->val.horodateur.heure,ptr->val.horodateur.min,ptr->val.horodateur.sec);
            printf("%-3c%-3d%d/%d/%-7d",ptr->val.sexe,ptr->val.age,ptr->val.Date.jour,ptr->val.Date.mois,ptr->val.Date.ans);
            printf("%-10s%-70s%-20s%-20s%-5d\n",ptr->val.period,ptr->val.repas,ptr->val.unite,ptr->val.lieu,ptr->val.nombre);
            i++;
        }
        ptr = ptr->next;
    }
    if(i==0)
        printf("Aucun element trouver... :(\n");
 }
//LA FONCTION clear_list SERT A EFFACER LA LISTE PASSEE EN PARAMETRE
 void clear_stack(list* li){
    list *ptr = li;
    while(li != NULL){
        li = li->next;
        free(ptr);
        ptr = li;
    }
 }
//LA FONCTION sort PERMET DE TRIER LES ELEMENTS D'UNE LISTE EN FONCTION DE L'AGE
void sort(list *li){
    list *ptr = li, *ctr;
    data temp;
    // le tri insertion //
    while(ptr->next != NULL){
        ctr = ptr->next;
        while(ctr != NULL){
            if(ptr->val.age > ctr->val.age){
                temp = ptr->val;
                ptr->val = ctr->val;
                ctr->val = temp;
            }
            ctr = ctr->next;
        }
        ptr = ptr->next;
    }
}
int main(){
    int choix,age;
    list *li;

    li = save_stack();
    do{
        printf("---------------------- MENU ------------------------\n");
        printf("0- Quiter\n");
        printf("1- Afficher la liste\n");
        printf("2- Rechercher et afficher les elements de la liste\n");
        printf("3- Trier la liste(en fonction de l'age)\n\n");
        printf("Faites un choix : ");
        scanf("%d",&choix);
        switch(choix){
            case 0: break;
            case 1: system("cls");
                    print_stack(li);
                    break;
            case 2: system("cls");
                    printf("Entrer l'age a rechercher : ");
                    scanf("%d",&age);
                    system("cls");
                    search(li,age);
                    break;
            case 3: system("cls");
                    sort(li);
                    printf("Le tri a ete effectuer avec succes :)\n");
                    break;
            default: printf("Faire un choix entre 0 et 3...\n");
        }
        system("PAUSE");
        system("cls");
    }while(choix!=0);
    clear_stack(li);

    return 0;
}

File: Road traffic data/Js/stack.js
//NOM: CHEYCHOU MOUAFO JUNIOR   mat: 21T2374
function save_tab(){
	var line = ["23/03/2022","16:26:22","M",21,"23/03/2022","12h-14h","Riz sauce d'arachide","Plateau","Maison",8];
	var lines = [];
	lines[0] = line;
	line = ["23/03/2022","16:27:24","F",18,"23/03/2022","14h-16h","Corn flakes","Bol","Maison",1];
	lines[lines.length] = line;
	line = ["23/03/2022","16:27:47","M",18,"12/01/2004","14h-16h","Riz suave arachide avec la viande","Plat","Maison",3];
	lines[lines.length] = line;
	line = ["23/03/2022","16:29:33","F",18,"23/03/2022","12h-14h","Purée d'avocat-pain","Plat","Maison",3];
	lines[lines.length] = line;
	line = ["23/03/2022","16:31:02","M",17,"23/03/2022","14h-16h","Rice and tomato stew","Plat","Maison",6];
	lines[lines.length] = line;
	line = ["23/03/2022","16:32:32","M",19,"23/03/2022","14h-16h","Riz sauce blanche","Plat","Maison",6];
	lines[lines.length] = line;
	line = ["23/03/2022","16:32:53","M",18,"23/03/2022","14h-16h","Riz blanc accompagné du haricot","Plat","Maison",7];
	lines[lines.length] = line;
	line = ["23/03/2022","16:33:37","M",22,"23/03/2022","12h-14h","Foufou ans Eru","Plat","Maison",3];
	lines[lines.length] = line;
	line = ["23/03/2022","16:33:52","M",22,"23/03/2022","12h-14h","Beignets haricots Bouilli","plat","Maison",2];
	lines[lines.length] = line;
	line = ["23/03/2022","16:34:02","M",19,"23/03/2022","12h-14h","Plantain frit et 2 oeufs","Assiette creuse","Maison",6];
	lines[lines.length] = line;
	line = ["23/03/2022","16:37:53","F",20,"23/03/2022","16h-18h","Les crêpes au lait","Bol","Maison",6];
	lines[lines.length] = line;
	line = ["23/03/2022","16:40:05","M",21,"14/04/2000","16h-18h","okok baton de manioc","Assiette creuse","Maison",5];
	lines[lines.length] = line;
	line = ["23/03/2022","16:41:40","M",18,"23/03/2022","10h-12h","Pain et une tasse de café au lait","Tasse de café","Maison",1];
	lines[lines.length] = line;
	line = ["23/03/2022","16:42:03","M",20,"22/03/2022","12h-14h","Le riz avec la sauce tomate","Plat","Maison",1];
	lines[lines.length] = line;
	line = ["23/03/2022","16:43:35","F",20,"23/03/2022","20h-22h","Sauce d'arrahide avec le manioc","Plat","Maison",6];
	lines[lines.length] = line;
	line = ["23/03/2022","16:43:40","F",17,"23/03/2022","16h-18h","Riz sauce d'arachide sans viande","Plat","Maison",5];
	lines[lines.length] = line;
	line = ["23/03/2022","16:43:53","F",20,"23/03/2022","14h-16h","Okok sucré ,manioc","Assiette creuse","Maison",6];
	lines[lines.length] = line;
	line = ["23/03/2022","16:45:54","M",36,"23/03/2022","12h-14h","Bouillon viande, plaintain mur","Assiette creuse","Restaurant",0];
	lines[lines.length] = line;
	line = ["23/03/2022","16:46:26","M",18,"23/03/2022","6h-10h","pain omelette","Plat","Route",0];
	lines[lines.length] = line;	
	
	return lines;
}

class node{
	constructor(elt){
		this.val = elt;
		this.next = null;
	}
}

class linked_list{
	constructor(){
		this.head = null;
		this.size = 0;
	}
	save(){
		var data = save_tab();
		var i = 0;
		while(i<19){
			var noeud = new node(data[i]);
			if(this.head == null){
				this.head = noeud;
				this.size++;
			}else{
				noeud.next = this.head;
				this.head = noeud;
				this.size++;
			}
			i++;
		}
	}
	
	afficher(){
		var pt = this.head;
		while(pt){
			console.log(pt.val[0]+'\t'+pt.val[1]+'\t'+pt.val[2]+'\t'+pt.val[3]+'\t'+pt.val[4]+'\t'+pt.val[5]+'\t'+pt.val[6]+'\t'+pt.val[7]+'\t'+pt.val[8]+'\t'+pt.val[9]+'\n');
			pt = pt.next;
		}
	}
	
	sort(){
		var j, i = 0;
		var ctr, pt = this.head;
		while(pt){
			ctr = pt.next;
			while(ctr){
				if(ctr.val[3] < pt.val[3]){
					var temp = ctr.val;
					ctr.val = pt.val;
					pt.val = temp;
				}
				ctr = ctr.next;
			}
			pt = pt.next;
		}
	}
	
	search(age){
		var i = 0;
		var pt = this.head;
		while(pt){
			if(age == pt.val[3]){
				i++;
				console.log(pt.val[0]+'\t'+pt.val[1]+'\t'+pt.val[2]+'\t'+pt.val[3]+'\t'+pt.val[4]+'\t'+pt.val[5]+'\t'+pt.val[6]+'\t'+pt.val[7]+'\t'+pt.val[8]+'\t'+pt.val[9]+'\n');
			}
			pt = pt.next;
		}
		if(i == 0)
			console.log('Aucun element trouver');
	}
}

var x = 1
stack = new linked_list();
while (x != 0){
    console.log("----------------------- MENU --------------------------");
    console.log("0 - Quitter");
    console.log("1 - Enregistrer les donnee");
    console.log("2 - Trier en fonction de l'age");
    console.log("3 - Rechercher les donnee en fonction de l'age\n");
    x = prompt("Faites un choix : ");
    x = parseInt(x);
    if (x == 1){
        stack.save();
		stack.afficher();
    }else if(x == 2){
        stack.sort();
		stack.afficher();
    }else if(x == 3){
        age = prompt("Entrer l'age a rechercher : ");
        age = parseInt(age);
        stack.search(age);
	}else if(x == 0){
		
    }else
        console.log("Faire un choix entre 0 et 3");
}
